src/main/java/org/jenkinsci/deprecatedusage/JenkinsFile.java

```diff
@@ -65,80 +52,92 @@ public String getWiki() {
+        if (Files.exists(file)) {
+            return CompletableFuture.supplyAsync(() -> {
+                if (checksum == null) {
+                    return false;
+                }
+                try {
+                    return checksum.matches(file);
+                } catch (IOException e) {
+                    System.out.println("Error validating checksum for " + file);
+                    System.out.println(e.toString());
+                    return false;
+                }
+            }, executor).thenCompose(checksumOk -> checksumOk ? CompletableFuture.completedFuture(null) : download(client));
+        }
+        return download(client);
+    }
+
+    private CompletableFuture<Void> download(CloseableHttpAsyncClient client) {
+        CompletableFuture<Void> future = new CompletableFuture<>();
+        try {
+            if (Files.exists(versionsRootDirectory)) {
+                Files.walkFileTree(versionsRootDirectory, new SimpleFileVisitor<Path>() {
+                    @Override
+                    public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) throws IOException {
+                        Files.delete(file);
+                        return FileVisitResult.CONTINUE;
+                    }
+
+                    @Override
+                    public FileVisitResult postVisitDirectory(Path dir, IOException exc) throws IOException {
+                        Files.delete(dir);
+                        return FileVisitResult.CONTINUE;
+                    }
+                });
+            }
+            Files.createDirectories(file.getParent());
+        } catch (IOException e) {
+            future.completeExceptionally(e);
+            return future;
         }
-        final String tmpPrefix = getClass().getPackage().getName() + '-';
-        downloadFuture = downloadExecutorService.submit(new Callable<Object>() {
+        client.execute(SimpleHttpRequests.get(url), new FutureCallback<SimpleHttpResponse>() {
             @Override
-            public Object call() throws IOException {
-                final File tempFile = File.createTempFile(tmpPrefix, ".tmp");
+            public void completed(SimpleHttpResponse result) {
                 try {
-                    final OutputStream output = new BufferedOutputStream(
-                            new FileOutputStream(tempFile));
-                    try {
-                        new HttpGet(url).copy(output);
-                    } finally {
-                        output.close();
+                    byte[] data = result.getBodyBytes();
+                    if (checksum == null || checksum.matches(data)) {
+                        Files.write(file, data, StandardOpenOption.CREATE, StandardOpenOption.TRUNCATE_EXISTING);
+                    } else {
+                        future.completeExceptionally(new DigestException("Invalid checksum for downloaded file " + file));
```

Comment:
```
Error downloading work/6ba2a29d/libvirt-slave/1.8.6/libvirt-slave.hpi
java.security.DigestException: Invalid checksum for downloaded file work/6ba2a29d/libvirt-slave/1.8.6/libvirt-slave.hpi
```

```
$ cat work/6ba2a29d/libvirt-slave/1.8.6/libvirt-slave.hpi
cat: work/6ba2a29d/libvirt-slave/1.8.6/libvirt-slave.hpi: No such file or directory
```

Output is at least misleading here.

Reply:
Damn relative directories! Yeah, fairly misleading since the script doesn't even tell you that it supports an optional CLI arg for specifying the output directory (defaults to creating a temp directory printed at the beginning). Will improve log messages.

Reply:
@jvz The problem here is that the update center advertises a file not yet on mirrors. It's a long standing infra issue. Still shouldn't complain about a checksum of a file that doesn't exist.

Reply:
I've improved the log messages a bit. Let me know if you still see anything weird.

Reply:
This still happens. When a file cannot be download (somewhat common due to INFRA-160), it still complains about an invalid checksum (which should never happen and is a serious problem).

Reply:
I added some manual code to retry a download when an invalid checksum is detected. I think this might be caused by a truncated download that doesn't cause a connection reset or other exception. The updated code will retry up to three times when invalid digests are found. This is a separate retry strategy from the HttpClient one.